src/wp-includes/post.php

```diff
@@ -4375,6 +4375,29 @@ function wp_publish_post( $post ) {
 		return;
 	}
 
+	// Ensure at least one term is applied for taxonomies with a default term.
+	foreach ( get_object_taxonomies( $post->post_type, 'object' ) as $taxonomy => $tax_object ) {
+		// Skip taxonomy if no default term is set.
+		if (
+			'category' !== $taxonomy &&
+			empty( $tax_object->default_term )
+		) {
+			continue;
+		}
+
+		// Do not modify previously set terms.
+		if ( ! empty( get_the_terms( $post, $taxonomy ) ) ) {
+			continue;
+		}
+
+		if ( 'category' === $taxonomy ) {
+			$default_term_id = (int) get_option( 'default_category' );
+		} else {
+			$default_term_id = (int) get_option( 'default_term_' . $taxonomy );
+		}
+		wp_set_post_terms( $post->ID, $default_term_id, $taxonomy );
```

Comment:
Should these be wrapped as an `array` and the `wp_set_post_terms` check skipped if it is empty? That would make sure the `! is_array( $tags )` branch of `wp_set_post_terms` doesn't get executed. I think that might introduce an issue with non-hierarchical taxonomies because the id will end up being cast to a string.

Reply:
I've made these changes in 34830cb, skipping in the event the option is not set and wrapping the term in an array for the function call.

Reply:
looks great to me!